93.md

```diff
@@ -0,0 +1,184 @@
+	tagBuilder.Grow(1000) // a guess
+	tagBuilder.WriteString(`[`)
+	for t, tag := range evt.Tags {
+		nItems := len(tag)
+		nsonIndex++
+		nsonBuf[nsonIndex] = uint8(nItems)
+
+		tagBuilder.WriteString(`[`)
+		for i, item := range tag {
+			v := strconv.Quote(item)
+			nsonIndex++
+			binary.BigEndian.PutUint16(nsonBuf[nsonIndex:], uint16(len(v)-2))
+			nsonIndex++
+			tagBuilder.WriteString(v)
+			if nItems > i+1 {
+				tagBuilder.WriteString(`,`)
+			}
+		}
+		tagBuilder.WriteString(`]`)
+		if nTags > t+1 {
+			tagBuilder.WriteString(`,`)
+		}
+	}
+	tagBuilder.WriteString(`]}`)
+	nsonBuf = nsonBuf[0 : nsonIndex+1]
+
+	kind := strconv.Itoa(evt.Kind)
+	kindChars := len(kind)
+	nsonBuf[0] = uint8(kindChars)
+
+	content := strconv.Quote(evt.Content)
+	contentChars := len(content) - 2
+	binary.BigEndian.PutUint16(nsonBuf[1:3], uint16(contentChars))
+
+	// actually build the json
+	base := strings.Builder{}
+	base.Grow(320 + // everything up to "nson":
+		2 + len(nsonBuf)*2 + // nson
+		9 + kindChars + // kind and its label
+		12 + contentChars + // content and its label
+		9 + tagBuilder.Len() + // tags and its label
+		2, // the end
+	)
+	base.WriteString(`{"id":"` + evt.ID + `","pubkey":"` + evt.PubKey + `","sig":"` + evt.Sig + `","created_at":` + strconv.FormatInt(int64(evt.CreatedAt), 10) + `,"nson":"`)
+	base.WriteString(hex.EncodeToString([]byte{uint8(len(nsonBuf) * 2)})) // nson size
+	base.WriteString(hex.EncodeToString(nsonBuf))                         // nson descriptors
+	base.WriteString(`","kind":` + kind + `,"content":` + content + `,"tags":`)
+	base.WriteString(tagBuilder.String() /* includes the end */)
+
+	return base.String()
+}
+```
+
+### Other restrictions
+
+Besides the field ordering and the presence of the `"nson"` field, other restrictions must be applied:
+
+- the `"created_at"` field must have 10, characters, which gives us a range of dates from about 20 years ago up to 250 years in the future.
+- to simplify decoding of `"content"` and `"tags"` strings, escape codes like `\uXXXX` are forbidden in NSON, UTF-8 must be used instead. Only `\n`, `\\` and `\"` are the only valid escaped sequences.
+
+### Backwards-compatibility
+
+Any reader who is not aware of the NSON-encoding can receive these events and decode them using whatever other JSON decoder they happen to have in hand. The `"nson"` field will just be ignored and life will continue as normal.
+
+Also, other event fields that may be present (for example, the NIP-03 `"ots"` field) can be added at the end, after `"tags"`, with no loss.
+
+### Other points worth mentioning
+
+- Relays can receive non-nsonified events from clients, then reformat and store them nsonified so they can serve future clients better by sending them NSON always.
+
+### Open questions to be edited out of the NIP
+
+- How to signal NSON support? I thought it would work to have an initial field `"n":1` (before `"id"`) on the JSON, which could be read very fast, but I don't know.
```

Comment:
Can it simply be signaled in the relay info document once its listed as supporting this nip?

Reply:
I think that's too restrictive, introduces the need for clients to query these relays and keep track of these settings in the code. Also I was thinking relays should be able to mix both nson and non-nson events (for whatever reason).

Another possibility is signaling on the websocket handshake, but that is also kinda annoying for everybody.

Reply:
Actually we don't need an initial field at all, the readers can just check if the static position `x...y` has the word `"nson"` in it.

Reply:
This would have false positives in the case of a non-nson even where the content field or a tag spans `x...y` and happens to have `nson` in those positions.

Reply:
yeah I think you should start with the signalling field of a fixed length to not screw the offsets. Then the variable length nson field after all the fixed length fields somewhere. I can't imagine how it could work otherwise.